enable checks

akshatagarwl opened the pull request.

akshatagarwl committed:
enable checks

cdepillabout commented on flake.nix:
```diff
@@ -25,18 +25,37 @@
         defaultPackage = packages.libspng;
 
-        # TODO checks
+        checks.libspng = pkgs.stdenv.mkDerivation {
+          pname = "libspng";
+          name = "libspng";
+          src = libspng-src;
+
+          doCheck = true;
+
+          mesonBuildType = "release";
+
+          mesonFlags = [
+            # this is required to enable testing
+            # https://github.com/randy408/libspng/blob/bc383951e9a6e04dbc0766f6737e873e0eedb40b/tests/README.md#testing
+            "-Ddev_build=true"
+          ];
+
+          outputs = [ "out" "dev" ];
+
+          checkInputs = with pkgs; [ cmake libpng ];
+          buildInputs = with pkgs; [ pkg-config zlib ];
+          nativeBuildInputs = with pkgs; [ ninja meson ];
+        };
```
I was thinking it might be a little cleaner to just override the `package`, since the derivation for running the tests is so similar:

```suggestion
        checks.libspng = packages.libspng.overrideAttrs (oldAttrs: {
          doCheck = true;

          mesonFlags = [
            # this is required to enable testing
            # https://github.com/randy408/libspng/blob/bc383951e9a6e04dbc0766f6737e873e0eedb40b/tests/README.md#testing
            "-Ddev_build=true"
          ];

          checkInputs = with pkgs; [ cmake libpng ];
        });
```

akshatagarwl replied:
This is great. Thanks!

cdepillabout commented on flake.nix:
```diff
@@ -25,18 +25,37 @@
         # `nix develop`
         devShell = pkgs.mkShell {
-          nativeBuildInputs = with pkgs; [ pkg-config zlib meson ninja ];
+          nativeBuildInputs = with pkgs; [ pkg-config zlib meson ninja cmake ];
```
You might be able to do a similar thing here, and just pull imports from the check defined above:

```suggestion
          inputsFrom = [ checks.libspng ];
```

Just like above, this would also help reduce duplication.

cdepillabout replied:
Oh, I didn't test that this actually gives you everything you need to compile the package.  You might want to do that.

(Although I did test that pkg-config, zlib, meson, ninja, cmake, and libpng were all available in the `nix develop` environment, so my guess is that it is fine.)

cdepillabout commented:
Oh, one more thing.  Could you add a small README.md to this repo explaining that this library is a replacement for `libpng`?  Probably also mention that it can't be directly used as it, but it can be used as a flake input.

akshatagarwl committed:
override the package in checks to remove code duplication

Since the derivation for building the package and the checks is very similar we can use override the package to make it a little cleaner

Co-authored-by: Dennis Gosnell <cdep.illabout@gmail.com>

akshatagarwl committed:
use inputsFrom in devShell to reduce duplication

Co-authored-by: Dennis Gosnell <cdep.illabout@gmail.com>

akshatagarwl commented:
@cdepillabout What does `inputsFrom` do? I couldn't find it documented anywhere

cdepillabout commented:
@humancalico `mkShell` has a little bit of documentation in the manual, but not enough: https://nixos.org/manual/nixpkgs/unstable/#sec-pkgs-mkShell

If you're looking for documentation I'd highly recommend you first look in the nixpkgs source code.  There is a lot of documentation hidden in comments in the source code that is not available anywhere on the web.  For instance, here's the docs for `mkShell` from the code (although in this case it is really no better than the nixpkgs manual):

https://github.com/NixOS/nixpkgs/blob/85568eb45cefe1838fe3f6d2eef8dcccdf04ceab/pkgs/build-support/mkshell/default.nix

If you wanted to send a PR to improve the documentation in either the nixpkgs manual or directly in this file, that would be a great contribution for Summer of Nix.

---------------------------------------------------------------------

As for your question about `inputsFrom`, `mkShell` takes the `nativeBuildInputs` and `buildInputs` from everything in the list to `inputsFrom`.

For instance, imagine you have a `mkShell` call like:

```nix
mkShell {
  inputsFrom = [ hello libspng ];
  nativeBuildInputs = [ spago ghc ];
  buildInputs = [zlib ];
}
```

This gets translated into something like:

```nix
mkShell {
  nativeBuildInputs = [ spago ghc ] ++ hello.nativeBuildInputs ++ libspng.nativeBuildInputs;
  buildInputs = [zlib ] ++ hello.buildInputs ++ libspng.buildInputs;
}
```

akshatagarwl commented:
Thanks @cdepillabout!
